worker/src/main/scala/com/lucidchart/piezo/ConnectionProvider.scala

```diff
@@ -1,25 +1,136 @@
+  private case class CachedIpWithExpiration(ip: String, expiration: Long)
+  // Cache for ip address and its expiration for a host
+  private val cachedIpWithExpiration: LinkedHashMap[String, CachedIpWithExpiration] = new LinkedHashMap()
+  // Time (in milliseconds) that the in-memory cache will retain the ip address
+  private val cachedIpTTL: Long = 1000
+  private val getIpNumRetries = 10
+
+  private val pool: Pool = new Pool(getIP)
+
+  // Intended to be used only for tests. This mocks an IP failover every time a connection is retreived
+  private val causeFailoverEveryConnection = if (dataSource != null) props.getProperty("org.quartz.dataSource." + dataSource + ".causeFailoverEveryConnection") == "true" else false
+
+  def createNewConnectionProvider(): Option[HikariCpPoolingConnectionProvider] = {
+    if(dataSource != null) {
+      Some(new HikariCpPoolingConnectionProvider(
+        props.getProperty("org.quartz.dataSource." + dataSource + ".driver"),
+        jdbcURL,
+        props.getProperty("org.quartz.dataSource." + dataSource + ".user"),
+        props.getProperty("org.quartz.dataSource." + dataSource + ".password"),
+        props.getProperty("org.quartz.dataSource." + dataSource + ".maxConnections").toInt,
+        props.getProperty("org.quartz.dataSource." + dataSource + ".validationQuery")
+      ))
+    } else {
+      logger.info("No job store found in config to get connections")
+      None
+    }
+  }
+
+  /**
+   * HikariCP connection pools don't automatically close when IP addresses for a hostname change. This function returns True, iff at
+   * least one of the following conditions is met:
+   *   - IP addresses have changed for the CNAME record used for DNS lookup
+   *   - causeFailoverEveryConnection is set to "true", which is used for testing failover functionality
+   *
+   * @param pool
+   *   the connection pool currently being used
+   * @param dnsIP
+   *   the IP returned when performing a DNS lookup
+   * @return
+   */
+  private def hasIpAddressChanged(pool: Pool, dnsIP: String): Boolean = {
+    causeFailoverEveryConnection || pool.ip != dnsIP
+  }
+
+  @tailrec
+  private def retryGettingIp(n: Int)(fn: => String): String = {
+    try {
+      return fn
+    } catch {
+      // Failed to resolve it from JVM
+      case e: UnknownHostException if n > 1 =>
+    }
+    // Wait 10 milliseconds between retries
+    Thread.sleep(10)
+    retryGettingIp(n - 1)(fn)
+  }
+
+  def _getIp: String = {
+    retryGettingIp(getIpNumRetries) {
+      // Get the ip address of the hostname. The result is cached in the JVM
+      val ip = java.net.InetAddress.getByName(dataSourceHostname).getHostAddress
+      cachedIpWithExpiration.put(dataSourceHostname, CachedIpWithExpiration(ip, System.currentTimeMillis() + cachedIpTTL))
+      ip
+    }
+  }
+
+  def getIP: String = {
+    val optionCachedValue = cachedIpWithExpiration.get(dataSourceHostname)
+    if (optionCachedValue.nonEmpty) {
```

Comment:
Why not do `optionCachedValue.map { cachedValue => ...` instead?

Reply:
Fair point. We use `.get` a lot in the repository already, but we can use a map here